cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -200,21 +184,53 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
   }
 
   @Override
+  public void dismissDataset(Dataset dataset) {
+    Preconditions.checkNotNull(dataset);
+    // static datasets cannot be dismissed
+    if (staticDatasets.containsValue(dataset)) {
+      LOG.warn("Attempt to dismiss static dataset {} from dataset cache", dataset);
+      return;
+    }
+    if (txContext == null || !(dataset instanceof TransactionAware)) {
+      dismissSafely(dataset);
```

Comment:
Yes rest of the logic is in the transaction context (DelayedDismissingTxContext). We need to hold on to the dataset until the transaction completes. Only the tx context knows when that happens. It overrides abort() and finish() to call dismissSafely() for all of the datasets that were marked for that using dismissAfterTx()